File: java/Exercises.java
import java.util.List;
import java.util.Map;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Optional;
import java.util.function.Predicate;
import java.io.BufferedReader;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;

public class Exercises {
    static Map<Integer, Long> change(long amount) {
        if (amount < 0) {
            throw new IllegalArgumentException("Amount cannot be negative");
        }
        var counts = new HashMap<Integer, Long>();
        for (var denomination : List.of(25, 10, 5, 1)) {
            counts.put(denomination, amount / denomination);
            amount %= denomination;
        }
        return counts;
    }

    public static Optional<String> firstThenLowerCase(List<String> strings, Predicate<String> condition) {
        return strings.stream()
                      .filter(condition)
                      .findFirst()
                      .map(String::toLowerCase);
    }

    public static SayPhrase say(String... words) {
        return new SayPhrase(String.join(" ", words));
    }

    public static class SayPhrase {
        private final String phrase;

        private SayPhrase(String phrase) {
            this.phrase = phrase;
        }

        public SayPhrase and(String word) {
            return new SayPhrase(this.phrase + " " + word);
        }

        public String phrase() {
            return this.phrase;
        }
    }

    static long meaningfulLineCount(String filename) throws IOException {
        try (var reader = new BufferedReader(new FileReader(filename))) {
             return reader.lines()
                .map(String::trim)
                .filter(line -> !line.isBlank() && !line.startsWith("#"))
                .count();
        }
    }
}

record Quaternion(double w, double x, double y, double z) {

    public static final Quaternion ZERO = new Quaternion(0, 0, 0, 0);
    public static final Quaternion I = new Quaternion(0, 1, 0, 0);
    public static final Quaternion J = new Quaternion(0, 0, 1, 0);
    public static final Quaternion K = new Quaternion(0, 0, 0, 1);

    public Quaternion {
        if (Double.isNaN(w) || Double.isNaN(x) || Double.isNaN(y) || Double.isNaN(z)) {
            throw new IllegalArgumentException("Coefficients cannot be NaN");
        }
    }

    public double a() {
        return w;
    }

    public double b() {
        return x;
    }

    public double c() {
        return y;
    }

    public double d() {
        return z;
    }

    public Quaternion plus(Quaternion q) {
        return new Quaternion(this.w + q.w, this.x + q.x, this.y + q.y, this.z + q.z);
    }

    public Quaternion times(Quaternion q) {
        return new Quaternion(
            this.w * q.w - this.x * q.x - this.y * q.y - this.z * q.z,
            this.w * q.x + this.x * q.w + this.y * q.z - this.z * q.y,
            this.w * q.y - this.x * q.z + this.y * q.w + this.z * q.x,
            this.w * q.z + this.x * q.y - this.y * q.x + this.z * q.w
        );
    }

    public Quaternion conjugate() {
        return new Quaternion(this.w, -this.x, -this.y, -this.z);
    }

    public List<Double> coefficients() {
        return Arrays.asList(w, x, y, z);
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
    
        if (w != 0.0) {
            sb.append(w);
        }
    
        if (x != 0.0) {
            if (sb.length() > 0 && x > 0) sb.append("+");
            if (x == 1.0) {
                sb.append("i");
            } else if (x == -1.0) {
                sb.append("-i");
            } else {
                sb.append(x).append("i");
            }
        }
    
        if (y != 0.0) {
            if (sb.length() > 0 && y > 0) sb.append("+");
            if (y == 1.0) {
                sb.append("j");
            } else if (y == -1.0) {
                sb.append("-j");
            } else {
                sb.append(y).append("j");
            }
        }

        if (z != 0.0) {
            if (sb.length() > 0 && z > 0) sb.append("+");
            if (z == 1.0) {
                sb.append("k");
            } else if (z == -1.0) {
                sb.append("-k");
            } else {
                sb.append(z).append("k");
            }
        }
        return sb.length() > 0 ? sb.toString() : "0";
    }
}


sealed interface BinarySearchTree permits Node, Empty {
    boolean contains(String nodeValue);
    BinarySearchTree insert(String nodeValue);
    int size();
}

final record Empty() implements BinarySearchTree {

    @Override
    public boolean contains(String nodeValue) {
        return false;
    }

    @Override
    public BinarySearchTree insert(String nodeValue) {
        return new Node(nodeValue, this, this);
    }

    @Override
    public int size() {
        return 0;
    }

    @Override
    public String toString() {
        return "()";
    }
}

final record Node(
String nodeValue, 
BinarySearchTree left, 
BinarySearchTree right) 
implements BinarySearchTree {

    @Override
    public boolean contains(String nodeValue) {
        if (this.nodeValue.equals(nodeValue)) {
            return true;
        }
        if (nodeValue.compareTo(this.nodeValue) < 0) {
            return left.contains(nodeValue);
        } else {
            return right.contains(nodeValue);
        }
    }

    @Override
    public BinarySearchTree insert(String nodeValue) {
        if (nodeValue.compareTo(this.nodeValue) < 0) {
            return new Node(this.nodeValue, left.insert(nodeValue), right);
        } else if (nodeValue.compareTo(this.nodeValue) > 0) {
            return new Node(this.nodeValue, left, right.insert(nodeValue));
        } else {
            return this;
        }
    }

    @Override
    public int size() {
        return 1 + left.size() + right.size();
    }

    @Override
    public String toString() {
        return ("(" + left + nodeValue + right + ")").replace("()", "");
    }
}    

File: kotlin/Exercises.kt
import java.io.BufferedReader
import java.io.FileReader
import java.io.IOException
import java.io.FileNotFoundException
import java.io.File


fun change(amount: Long): Map<Int, Long> {
    require(amount >= 0) { "Amount cannot be negative" }
    
    val counts = mutableMapOf<Int, Long>()
    var remaining = amount
    for (denomination in listOf(25, 10, 5, 1)) {
        counts[denomination] = remaining / denomination
        remaining %= denomination
    }
    return counts
}


fun firstThenLowerCase(strings: List<String>, predicate: (String) -> Boolean): String? {
    return strings.firstOrNull(predicate)?.lowercase()
}


class Say(private val words: List<String> = emptyList()) {

    fun and(word: String): Say {
        return Say(words + word)
    }

    val phrase: String
        get() = words.joinToString(" ")
}

fun say(word: String = ""): Say {
    return if (word.isEmpty()) {
        Say()
    } else {
        Say(listOf(word))
    }
}


fun meaningfulLineCount(filename: String): Long {
    if (!File(filename).exists()) {
        throw FileNotFoundException("No such file")
    }
    return BufferedReader(FileReader(filename)).use { reader ->
        reader.lineSequence()
            .filter { it.isNotBlank() && !it.trimStart().startsWith("#") }
            .count().toLong()
    }
}


data class Quaternion(val w: Double, val x: Double, val y: Double, val z: Double) {

    companion object {
        val ZERO = Quaternion(0.0, 0.0, 0.0, 0.0)
        val I = Quaternion(0.0, 1.0, 0.0, 0.0)
        val J = Quaternion(0.0, 0.0, 1.0, 0.0)
        val K = Quaternion(0.0, 0.0, 0.0, 1.0)
    }

    val a: Double
        get() = w
    val b: Double
        get() = x
    val c: Double
        get() = y
    val d: Double
        get() = z

    operator fun plus(q: Quaternion): Quaternion {
        return Quaternion(w + q.w, x + q.x, y + q.y, z + q.z)
    }

    operator fun times(q: Quaternion): Quaternion {
        return Quaternion(
            w * q.w - x * q.x - y * q.y - z * q.z,
            w * q.x + x * q.w + y * q.z - z * q.y,
            w * q.y - x * q.z + y * q.w + z * q.x,
            w * q.z + x * q.y - y * q.x + z * q.w
        )
    }

    fun conjugate(): Quaternion {
        return Quaternion(w, -x, -y, -z)
    }

    fun coefficients(): List<Double> {
        return listOf(w, x, y, z)
    }

    override fun toString(): String {
        val sb = StringBuilder()
        if (w != 0.0) sb.append(w)
        if (x != 0.0) {
            if (sb.isNotEmpty() && x > 0) sb.append("+")
            if (x == -1.0) sb.append("-i") else if (x == 1.0) sb.append("i") else sb.append(x).append("i")
        }
        if (y != 0.0) {
            if (sb.isNotEmpty() && y > 0) sb.append("+")
            if (y == -1.0) sb.append("-j") else if (y == 1.0) sb.append("j") else sb.append(y).append("j")
        }
        if (z != 0.0) {
            if (sb.isNotEmpty() && z > 0) sb.append("+")
            if (z == -1.0) sb.append("-k") else if (z == 1.0) sb.append("k") else sb.append(z).append("k")
        }
        return if (sb.isEmpty()) "0" else sb.toString()
    }
}


sealed interface BinarySearchTree {
    fun contains(nodeValue: String): Boolean
    fun insert(nodeValue: String): BinarySearchTree
    fun size(): Int

    object Empty : BinarySearchTree {
        override fun contains(nodeValue: String): Boolean = false
        override fun insert(nodeValue: String): BinarySearchTree = Node(nodeValue, Empty, Empty)
        override fun size(): Int = 0
        override fun toString(): String = "()"
    }

    data class Node
    (
    val nodeValue: String, 
    val left: BinarySearchTree, 
    val right: BinarySearchTree
    ) : 
    BinarySearchTree {
        override fun contains(nodeValue: String): Boolean {
            return when {
                this.nodeValue == nodeValue -> true
                nodeValue < this.nodeValue -> left.contains(nodeValue)
                else -> right.contains(nodeValue)
            }
        }

        override fun insert(nodeValue: String): BinarySearchTree {
            return when {
                nodeValue < this.nodeValue -> copy(left = left.insert(nodeValue))
                nodeValue > this.nodeValue -> copy(right = right.insert(nodeValue))
                else -> this
            }
        }

        override fun size(): Int = 1 + left.size() + right.size()

        override fun toString(): String = "($left$nodeValue$right)".replace("()", "")
    }
}